File: Data Structures & Algorithms/binary-tree-diameter/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int best = 0;
        depth(root, best);
        return best;
    }

private:
    int depth(TreeNode* node, int& best) {
        if (!node) return 0;
        int left  = depth(node->left, best);
        int right = depth(node->right, best);
        best = max(best, left + right);   // path bending at this node
        return 1 + max(left, right);      // height, for the parent
    }
};

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        left, right = self.findHeight(root.left), self.findHeight(root.right)
        if abs(left - right) > 1:
            return False
        return (self.isBalanced(root.left) and self.isBalanced(root.right))
        
    def findHeight(self, node):
        if not node:
            return 0
        left, right = self.findHeight(node.left), self.findHeight(node.right)
        return 1 + max(left, right)